File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int t = Integer.parseInt(br.readLine());

        for (int i = 0; i < t; i++) {
            String str = br.readLine();
            Stack<Character> stack = new Stack<>();
            for (int j = 0; j < str.length(); j++) {
                if(!stack.isEmpty() && stack.peek() == '(' && str.charAt(j) == ')')
                    stack.pop();
                else
                    stack.push(str.charAt(j));
            }
            if(stack.isEmpty()) System.out.println("YES");
            else System.out.println("NO");
        }
    }
}

File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline

t = int(input())

for _ in range(t):
    command = input().rstrip()
    stack = []
    
    for i in command:
        if len(stack) == 0 and i == '(':
            stack.append(i)
        elif len(stack) == 0 and i == ')':
            stack.append(i)
        elif stack[-1] == '(' and i == ')':
            stack.pop()
        else:
            stack.append(i)
    
    if len(stack) == 0:
        print('YES')
    else:
        print('NO')